### ISSUE TITLE
Function erroneously returns null when compiled in release mode

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/coreclr#18332]
@leecow commented on [Wed Jun 06 2018](https://github.com/dotnet/corefx/issues/30173)

@jlwSL commented on [Wed Jun 06 2018](https://github.com/dotnet/core/issues/1663)

I've discovered a bug that causes a function to return null even if every possible return value is non-null. The issue does not happen if the project is compiled in debug mode. Please see the code attached that reproduces this issue.

```csharp
using System;
using System.Collections.Generic;

namespace DotNetBug
{
    internal class Foo : IDisposable
    {
        public void Dispose()
        {
        }
    }

    class Program
    {
        public static string Aargh()
        {
            using (var foo = new Foo())
            {
                foreach (var i in new List<int>())
                {
                    try
                    {
                        Console.WriteLine("here"); // This doesn't get printed
                    }
                    catch (Exception)
                    {
                        return "wibble";
                    }
                }
                foreach (var i in new List<int>())
                {
                }
            }
        
            return "wobble";
        }

        static void Main(string[] args)
        {
            Console.WriteLine(Aargh() == null);
        }
    }
}
```
The code above prints True.

For your information, my system details are below:

```bash
$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   2.1.300
 Commit:    adab45bf0c

Runtime Environment:
 OS Name:     centos
 OS Version:  7
 OS Platform: Linux
 RID:         centos.7-x64
 Base Path:   /usr/share/dotnet/sdk/2.1.300/

Host (useful for support):
  Version: 2.1.0
  Commit:  caa7b7e2ba

.NET Core SDKs installed:
  2.1.300 [/usr/share/dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

I can also reproduce the issue on Windows with a slightly older SDK and Host version:
```bash
$ dotnet --info
.NET Command Line Tools (2.1.201)

Product Information:
 Version:            2.1.201
 Commit SHA-1 hash:  7932dc6179

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17134
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.1.201\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.7
  Build    : 2d61d0b043915bc948ebf98836fefe9ba942be11
```

I tried to reproduce the issue targeting .NET Framework 4.5 and 4.6.1 and it worked correctly in both of those cases.

---

@leecow commented on [Wed Jun 06 2018](https://github.com/dotnet/core/issues/1663#issuecomment-395163136)

I ran your code on Mac, Fedora 28 and a clean CentOS 7 image and all return `false`. Going to move this to the CoreFX to start. 

---

@0xd4d commented on [Wed Jun 06 2018](https://github.com/dotnet/corefx/issues/30173#issuecomment-395183992)

Looks like a jitter bug, so coreclr is a better repo.

> I ran your code on Mac, Fedora 28 and a clean CentOS 7 image and all return false

Did you try Release builds?

**Comment by danmosemsft** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
Disabling finally cloning "fixes" things. Not sure if that's actually the bug, so let me look deeper.

**Comment by AndyAyersMS** [comment]
Yes, this looks like a bug in finally cloning.

If a try-finally has multiple callfinallies they will appear in some sequence in the flow graph after the last block of the try. And sometimes these callfinallies are associated with different finally clauses (as is the case here).

Schematically
```
   LastTryBlock;
   CallFinally1; --> Finally A
   GotoContinuation1;
   CallFinally2; --> Finally B
   GotoContinuation2;
   ...
   Finally A: nop; retfinally;
   Finally B: nop; retfinally;
   ...
   Continuation1: return "wobble";
   Continuation2: return loc2;
```
   
If the jit decides to clone one of the finallies and the corresponding callfinally is not the first one in the callfinally chain, the jit will first try and move the corresponding callfinally to just after the end of the try region, with the hope that once cloning is done, control flow can now just fall out of the try and into the clone.  So here the jit reorders:
```
   LastTryBlock;
   CallFinally2; --> Finally B
   GotoContinuation2;
   CallFinally1; --> Finally A
   GotoContinuation1;
   ...
   Finally A: nop; retfinally;
   Finally B: nop; retfinally;
   ...
   Continuation1: return "wobble";
   Continuation2: return loc2;
```
and then clones:
```
   LastTryBlock;
   FinallyBClone: nop; goto Continuation2;
   CallFinally1; --> Finally A
   GotoContinuation1;
   ...
   Finally A: nop; retfinally;
   Finally B: nop; retfinally;
   ...
   Continuation1: return "wobble";
   Continuation2: return loc2;
```
But this block rearrangement was missing an important check. The jit must also ensure that `LastTryBlock` targets the callfinally that the jit plans to move. Absent this check, finally cloning may cause the wrong finally to be invoked upon normal exit from a try-finally.

That's what happens here.  The finally `B` associated with the catch clause is executed instead finally `A` leading to the final return. And so the jit ends up returning the zero-inited `loc2`and hence null.

**Comment by AndyAyersMS** [comment]
Thanks for reporting this, @jlwSL.

Any objection to me using your code as a test case here in the repo? Also, I'm curious how you found this....

**Comment by AndyAyersMS** [comment]
Note this particular optimization is not enabled on desktop. So this is a core-only issue. 

Finally cloning was introduced in #8551. The bug should repro in 2.0 as well.

**Comment by jlwSL** [comment]
@AndyAyersMS I'm very happy for you to use the code as a test case.

I hit the code in production. We recently switched over from using Mono to .NET Core and one of our testers noticed some odd behaviour and it boiled down to returning null when it wasn't possible to. I stripped away as much code as I could so that I could provide a code sample that would allow you to reproduce the bug.

How do you propose I proceed from here? I would really like to keep using .NET Core. Is there any way I can compile in release mode but disable this optimisation? Would it be possible to get .NET Core 2.1 patched before waiting for .NET Core 3?

**Comment by AndyAyersMS** [comment]
If you have this isolated down to just one or a small number of methods, you can selectively disable optimization for those methods only, via `[MethodImpl(MethodImplOptions.NoOptimization)]`.  Note this disables all optimizations -- there is no way to just turn this one off selectively.

If this pattern appears throughout your code then you'll have to wait for a fix. I think this is a good candidate for inclusion in 2.1 servicing.  Should have the fix ready today for master.

**Comment by jlwSL** [comment]
Thanks @AndyAyersMS. I'm afraid I don't understand your reasoning for why this bug occurred so could you please tell me what patterns to look out for in my code to spot this bug? I also noticed that putting something as simple as a Console.WriteLine before the using end brace was enough to not hit this bug. Is this a reliable workaround?

**Comment by AndyAyersMS** [comment]
It is a bit tricky to describe since the jit operates at an IL level -- so we have to back-map the patterns we see in IL to the C# that might inspire CSC to produce such IL. Unfortunately this is not always easy.

To trigger this issue, we first of all need a nested set of `try/finally`s. These can be explicitly coded in source or can appear implicitly when you use `using` or `fixed` or `foreach` blocks (and possibly other constructs). So in your example above the `using` creates a `try/finally` and nested inside the `try` are two more `try/finally`s from the `foreach`s.

Second, one of the nested `try/finally`s needs to appear right at the end of the outer `try/finally`. Here that second `foreach` loop is right at the end of the `using` block.

Third, that last nested `try/finally` must have an empty finally clause. In your case the `foreach` inspired `try/finally` is there to ensure that `Dispose` is called on the `List<int>` enumerator. The jit inlines this call and sees that `List<int>.Dispose` doesn't do anything. So that finally block is empty and that nested `try/finally` can be removed. This removal is what actually lays the groundwork for the bug.

Fourth, the outer `try/finally` must have two distinct exit paths. Here there is one path leading to print `wobble` (the normal exit path) and another leading to print `wibble` (an exceptional exit path).

Fifth, the jit must decide to clone the finally that is not the normal exit path. This is probably not something I can easily map back to C#.

So roughly speaking, we need to see source code with the following shape:
```C#
try   // from using in your case
{
      // arbitrary stuff can go here
      // needs to be a return in here somehwere

              ... return "a";

     try   // from foreach in your case
     {
           // arbitrary stuff can go here
     }
     finally
     {
           // finally must end up empty 
     }
     // nothing can go here (see note below)
}
finally
{
      // arbitrary stuff can go here
}

return "b";
```
If you look at the testing I did for the fix, roughly 20 out of 300,000 methods I tested had this pattern, so it's not very common in the code we use to test the jit. But it may be more common in your code base.

Because the jit won't inline methods with `try/finally` constructs, the `try/finally` nesting must appear all in one method. But spotting this in source can be tricky because of implicit use of `try/finally`.

When you add a call at the end of the `using` it breaks the pattern as that call ends up n the `nothing can go here` spot. So yes it is a workaround, provided that the call is not inlined, or if inlined, does something that can't be optimized away.

**Comment by AndyAyersMS** [comment]
Re-opening to track port of fix to 2.1.

**Comment by jlwSL** [comment]
Thank you for the detailed explanation @AndyAyersMS  and thank you for reacting so quickly to this bug report. You and everyone working on .NET Core are doing such a fantastic job!

**Comment by AndyAyersMS** [comment]
Fixed in 2.1 servicing via #18384.
